kratos/includes/registry.h

```diff
@@ -0,0 +1,275 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
+//
+//  Main authors:    Pooyan Dadvand
+//
+
+#if !defined(KRATOS_REGISTRY_H_INCLUDED )
+#define  KRATOS_REGISTRY_H_INCLUDED
+
+
+// System includes
+#include <string>
+#include <iostream>
+
+
+// External includes
+
+
+// Project includes
+#include "includes/registry_item.h"
+#include "includes/registry_value_item.h"
+#include "utilities/parallel_utilities.h"
+#include "utilities/string_utilities.h"
+
+namespace Kratos
+{
+///@addtogroup ApplicationNameApplication
+///@{
+
+///@name Kratos Globals
+///@{
+
+///@}
+///@name Type Definitions
+///@{
+
+///@}
+///@name  Enum's
+///@{
+
+///@}
+///@name  Functions
+///@{
+
+///@}
+///@name Kratos Classes
+///@{
+
+/// Short class definition.
+/** Detail class definition.
+*/
+class Registry
+{
+public:
+    ///@name Type Definitions
+    ///@{
+
+    /// Pointer definition of Registry
+    KRATOS_CLASS_POINTER_DEFINITION(Registry);
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /// Default constructor.
+    Registry(){}
+
+    /// Destructor.
+    virtual ~Registry(){}
+
+    ///@}
+    ///@name Operators
+    ///@{
+
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+
+    template< typename TItemType, class... TArgumentsList >
+    static RegistryItem& AddItem(std::string const& ItemFullName, TArgumentsList&&... Arguments){
+
+        const std::lock_guard<LockObject> scope_lock(ParallelUtilities::GetGlobalLock());
+
+        auto item_path = StringUtilities::SplitStringByDelimiter(ItemFullName, '.');
+        KRATOS_ERROR_IF(item_path.empty()) << "The item full name is empty" << std::endl;
+
+        RegistryItem* p_current_item = &GetRootRegistryItem();
+
+        for(std::size_t i = 0 ; i < item_path.size() - 1 ; i++){
+            auto& item_name = item_path[i];
+            if(p_current_item->HasItem(item_name)){
+                p_current_item = &p_current_item->GetItem(item_name);
+            }
+            else{
+                p_current_item = &p_current_item->AddItem<RegistryItem>(item_name);
+            }
+        }
+
+        // I am doing the last one out of the loop to create it with the given type and argument
```

Comment:
What happens is someone tries to add a "leave" with the same name as a path_item inserted before but which is not of the `RegistryItem` type?

Example:
```
Registry::AddItem<RegistryValueItem<double>>("path.to.the.registry.new_item");
Registry::AddItem<RegistryValueItem<double>>("path.to.the.registry");
```

Reply:
It woks.

There is a test for this.

Reply:
Ok, but then if you do it the other way around, it will work, because a `RegistryValueItem<T>` is a `RegistryItem` no?

Reply:
Good question! It will work because I don't check the type of the father and RegistryValue can also work as RegistryItem, so there won't be an issue here. In fact, I am checking and sending an error if it exists to avoid such situations in which you think you add one type of Item but another type already exists there and the behavior is not what you expected.